### DISCUSSION TITLE
Generic type constraint: Class<T> where T : null

### DISCUSSION THREAD

**Discussion by jez9999** [dotnet/csharplang#6726] [Language Ideas]
Right now you can declare a class with a generic type constraint guaranteeing that `T` is not nullable.  But I want to do the opposite.  I want to only allow types that are nullable.  As far as I can tell that doesn't exist right now.  Outside of a 'nullable enable context' this would require `T` to be a reference type.  Inside a 'nullable enable context' this would require `T` to be a nullable.

1) Can this be achieved right now in C#?
2) If not, could it be added?

**Comment by TahirAhmadov** [comment]
Curious, what is the use case for this?

**Comment by jez9999** [comment]
A dictionary class where the indexer can return null when the key doesn't exist rather than throwing an exception.  Sure, you can use `TryGetValue` but it's not as nice and compact as the indexing syntax.  Of course, this means `TValue` needs to be able to be null.

**Reply by TahirAhmadov** [reply]
This is how I solve that:
```cs
public static TValue? GetOrNull<TKey, TValue>(this IDictionary<TKey, TValue> This, TKey key)
  where TKey : notnull where TValue : class
  => This.TryGetValue(key, out var res) ? res : null;

public static TValue? GetOrNullStruct<TKey, TValue>(this IDictionary<TKey, TValue> This, TKey key)
  where TKey : notnull where TValue : struct
  => This.TryGetValue(key, out var res) ? res : null;

// elsewhere
string? strValue = stringDict.GetOrNull("key");
int? intValue = intDict.GetOrNullStruct("other key");
```

**Reply by jez9999** [reply]
Not the indexer syntax though, which would be nice.  I think I'm used to that behaviour in Perl or JS or something.  :-)

**Reply by TahirAhmadov** [reply]
If something is done a certain way in Perl or JS, it's a very strong reason not to do it like that in C# :)

**Reply by jez9999** [reply]
I was waiting for that wise crack!  But not everything they do is bad, frankly.

**Reply by TonyValenti** [reply]
I do something similar to @TahirAhmadov but it has the side effect that I can't use my method in a generic context because I actually have two different methods. Which I hate.

**Comment by gurustron** [comment]
See [this](https://github.com/dotnet/csharplang/discussions/6931#discussioncomment-4830614) answer - it seems currently it is impossible to achieve that due to the difference in how NRTs and nullable value types are implemented.

**Reply by TahirAhmadov** [reply]
Having considered all the inputs, it seems that the only way to achieve this in a seamless fashion is the "multiple versions of IL per un-constrained generic argument". Needless to say, this would be a monumentally expensive undertaking. While I understand why the runtime and C# teams were unwilling to attempt to achieve this, I still wish they did. :) C'est la vie

**Reply by AartBluestoke** [reply]
could this be done with only 2? a value Type, and an object Type?
